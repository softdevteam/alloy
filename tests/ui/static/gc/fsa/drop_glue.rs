#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
include!{"./auxiliary/types.rs"}

impl<T: Debug> Drop for Wrapper<T> {
    fn drop(&mut self) {
        use_val(&self.0);
    }
}

impl<T: Debug> Drop for FinalizerUnsafeWrapper<T> {
    fn drop(&mut self) {
        use_val(&self.0);
    }
}

fn main() {
    Gc::new(Wrapper(FinalizerUnsafeWrapper(FinalizerUnsafeWrapper(FinalizerUnsafeType(1)))));
    //~^ ERROR: The drop method for `Wrapper<FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>>` cannot be safely finalized.
    //~| ERROR: The drop method for `FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>` cannot be safely finalized.
}
