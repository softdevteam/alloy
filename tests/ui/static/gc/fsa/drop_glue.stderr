error: The drop method for `Wrapper<FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>>` cannot be safely finalized.
  --> $DIR/drop_glue.rs:19:13
   |
LL |         use_val(&self.0);
   |                 -------
   |                 |
   |                 a finalizer cannot safely use this `FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>`
   |                 from a drop method because it does not implement `Send` + `Sync`.
...
LL |     Gc::new(Wrapper(FinalizerUnsafeWrapper(FinalizerUnsafeWrapper(FinalizerUnsafeType(1)))));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<Wrapper<FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>>>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values which are thread-safe.

error: The drop method for `FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>` cannot be safely finalized.
  --> $DIR/drop_glue.rs:19:13
   |
LL |         use_val(&self.0);
   |                 -------
   |                 |
   |                 a finalizer cannot safely use this `FinalizerUnsafeWrapper<FinalizerUnsafeType>`
   |                 from a drop method because it does not implement `Send` + `Sync`.
...
LL |     Gc::new(Wrapper(FinalizerUnsafeWrapper(FinalizerUnsafeWrapper(FinalizerUnsafeType(1)))));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<Wrapper<FinalizerUnsafeWrapper<FinalizerUnsafeWrapper<FinalizerUnsafeType>>>>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values which are thread-safe.

error: aborting due to 2 previous errors

